web-api/src/storage.rs

```diff
@@ -119,29 +119,33 @@ pub(crate) trait DocumentProperty {
 #[async_trait]
 pub(crate) trait Interest {
     async fn get(&self, user_id: &UserId) -> Result<UserInterests, Error>;
+}
 
-    async fn update_positive<F>(
-        &self,
-        doc_id: &DocumentId,
-        user_id: &UserId,
-        update_cois: F,
-    ) -> Result<(), Error>
-    where
-        F: Fn(&mut Vec<PositiveCoi>) -> &PositiveCoi + Send + Sync;
+pub(crate) struct InteractionUpdateContext<'a, 'b> {
+    pub(crate) document: &'a PersonalizedDocument,
+    pub(crate) category_weight_diff: &'a mut i32,
+    pub(crate) positive_cois: &'b mut Vec<PositiveCoi>,
 }
 
 #[async_trait]
 pub(crate) trait Interaction {
     async fn get(&self, user_id: &UserId) -> Result<Vec<DocumentId>, Error>;
 
     async fn user_seen(&self, id: &UserId) -> Result<(), Error>;
+
+    async fn update_interactions<F>(
```

Comment:
The method updates (our knowledge about) the interaction a user did, that this mainly updates interests is an implementation detail.

Through I would argue we don't have much benefit from splitting the database methods by that many traits when the implementations can't really exist independent of each other and we also don't have any use case for partially implementing some of them. 

I did something similar initially in the discovery engine to get some scoping and use it for some transaction things (which didn't work out and never ended up in the code). But we did remove the scoping, we don't have really separate concerns, they are all implemented on the same type, they have overlapping concerns (like transactions) and they are not split by db-backend or similar either as this won't work for e.g. `Documents::get_by_ids`. 

> Probably the best solution will be to have an abstraction for transactions.

I agree, but that's out side of the scope of the PR I think. I intentionally factored out all the db interactions into functions so that it will be faster to swap over to using a transactions abstraction.

Reply:
:pushpin:  I split that out into a separate PR as it involved to many unrelated changes once I started doing it.